src/environment.js

```diff
@@ -0,0 +1,1118 @@
+        var result = [];
+        if (json != null) {
+            result = json.translations || [];
+        }
+        return result;
+    }
+};
+
+class Environment {
+    // Intialize an environment object that will be used to render
+    // all of the macros in one document or page. We pass in a context
+    // object (which may come from HTTP request headers) that gives
+    // details like the page title and URL. These are available to macros
+    // through the global 'env' object, and some of the properties
+    // are also copied onto the global 'page' object.
+    //
+    // Note that we don't use the Environment object directly when
+    // executing macros. Instead call getExecutionContext(), supplying
+    // the macro arguments list to get an object specific for executing
+    // one macro.
+    //
+    // Note that we pass the Templates object when we create an Environment.
+    // this is so that macros can recursively execute other named macros
+    // in the same environment.
+    //
+    // The optional third argument is for use only by tests. Setting it to
+    // true makes us not freeze the environment so that tests can stub out
+    // methods in the API like mdn.fetchJSONResources
+    //
+    constructor(perPageContext, templates, testing = false) {
+        // Freeze an object unless we're in testing mode
+        function freeze(o) {
+            return testing ? o : Object.freeze(o);
+        }
+
+        /**
+         * For each function-valued property in o, bind that function
+         * to the specified bindings object, if there is one. Also,
+         * create lowercase and titlecase variants of each property in
+         * o, and finally, freeze the object o and return it.
+         *
+         * The binding means that the functions defined in this file
+         * can use `this.__kumascript` to refer to the global
+         * kumascript environment so that they can do
+         * `this.__kumascript.env.locale` for example.
+         *
+         * The case-insensitive variants implement legacy behavior in
+         * KumaScript, where macros can use case-insensitive names of
+         * objects and methods.
+         *
+         * And the freeze() call is a safety measure to prevent
+         * macros from modifying the execution environment.
+         */
+        function prepareProto(o, binding) {
+            let p = {};
+            for (let [key, value] of Object.entries(o)) {
+                if (binding && typeof value === 'function') {
+                    value = value.bind(binding);
+                }
+                p[key] = value;
+                p[key.toLowerCase()] = value;
+                p[key[0].toUpperCase() + key.slice(1)] = value;
+            }
+            return freeze(p);
+        }
+
+        this.templates = templates;
+        let globals = Object.create(prepareProto(globalsPrototype));
+        globals.__kumascript = globals;
```

Comment:
I don't understand the need for this. It seems that within the functions that have been bound to `globals` by `prepareProto`, it should be equally valid to use, for example, `this.env.locale` as `this.__kumascript.env.locale`. What am I missing?

Reply:
Fixed in 39dac67

While fixing up this file, it occurred to me that I could move the big blobs of API functions (mdnPrototype, wikiPrototype, stringPrototype, etc.) into their own modules, maybe in a new  src/api/ directory. Then I could require() those into environment.js. It would make the file not so long, and would separate the code with poor test coverage from the code with good test coverage.

I like the idea, and have added a TODO comment about it. With your approval, I'll go ahead and make that change as part of this PR. Though if you'd rather not have a change that large at this stage of the review, I'll hold off.